### PR TITLE
Validator API

### PR DESCRIPTION
**Pull request by spble**
## Proposed Changes

This PR extends the REST API endpoints, to replace the validator gRPC interface with a REST interface.

This PR lays the foundations for some of the API endpoints, and starts with the 'validator duties' function.

More functions will be added in a future PR

### CODE REVIEW

File: beacon_node/rest_api/src/validator.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use store::Store;
+use types::beacon_state::EthSpec;
+use types::{BeaconBlock, BeaconState, Epoch, RelativeEpoch, Shard, Slot};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ValidatorDuty {
+    /// The validator's BLS public key, uniquely identifying them. _48-bytes, hex encoded with 0x prefix, case insensitive._
+    #[serde(rename = "validator_pubkey")]
```
Comment (paulhauner) [inline beacon_node/rest_api/src/validator.rs]: What's the deal with these `renames`? Does serde to `validator-pubkey`?

File: beacon_node/rest_api/src/validator.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            .map(|pk| parse_pubkey(pk))
+            .collect::<Result<Vec<_>, _>>()?,
+        Err(e) => {
+            return Err(e);
+        }
+    };
+    let mut duties: Vec<ValidatorDuty> = Vec::new();
+
+    // Get a list of all validators for this epoch
+    let validator_proposers: Result<Vec<usize>, _> = epoch
```
Comment (paulhauner) [inline beacon_node/rest_api/src/validator.rs]: I think lines 81..93 could become:

```rust
    // Get a list of all validators for this epoch
    let validator_proposers: Result<Vec<usize>, _> = epoch
        .slot_iter(T::EthSpec::slots_per_epoch())
        .map(|slot| head_state.get_beacon_proposer_index(slot, relative_epoch, &beacon_chain.spec))
        .collect()
        .map_err(|e| ApiError:: ... )?;
```

File: beacon_node/rest_api/src/helpers.rs [hunk 3]
```diff
@@ -31,6 +36,24 @@ pub fn parse_root(string: &str) -> Result<Hash256, ApiError> {
     }
 }
 
+/// Parse a PublicKey from a `0x` prefixed hex string
+pub fn parse_pubkey(string: &str) -> Result<PublicKey, ApiError> {
```
Comment (paulhauner) [inline beacon_node/rest_api/src/helpers.rs]: I think this function permits both with and without the `0x` prefix. I would have thought the `0x` prefix is mandatory.

File: beacon_node/rest_api/src/validator.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        Ok((_, v)) => Epoch::new(v.parse::<u64>().map_err(|e| {
+            ApiError::InvalidQueryParams(format!("Invalid epoch parameter, must be a u64. {:?}", e))
+        })?),
+        Err(_) => {
+            // epoch not supplied, use the current epoch
+            current_epoch
+        }
+    };
+    let relative_epoch = RelativeEpoch::from_epoch(current_epoch, epoch)
+        .map_err(|e| ApiError::InvalidQueryParams(format!("Cannot get RelativeEpoch: {:?}", e)))?;
```
Comment (paulhauner) [inline beacon_node/rest_api/src/validator.rs]: We'll get this error if `epoch` is more than `1` epoch from `head_state`.

We could perhaps return an error message like "Epoch out of range"

File: beacon_node/rest_api/src/validator.rs (hunk 1 above)
Comment (spble) [reply beacon_node/rest_api/src/validator.rs]: No, these were automatically generated from the OpenAPI

Comment (spble) [reply beacon_node/rest_api/src/validator.rs]: They aren't necessary and I will remove them.

File: beacon_node/rest_api/src/validator.rs (hunk 2 above)
Comment (spble) [reply beacon_node/rest_api/src/validator.rs]: Cool, thanks. I've reduced it to something, similar to your suggestion:
```
// Get a list of all validators for this epoch
    let validator_proposers: Vec<usize> = epoch
        .slot_iter(T::EthSpec::slots_per_epoch())
        .map(|slot| {
            head_state
                .get_beacon_proposer_index(slot, relative_epoch, &beacon_chain.spec)
                .map_err(|e| {
                    ApiError::ServerError(format!(
                        "Unable to get proposer index for validator: {:?}",
                        e
                    ))
                })
        })
        .collect::<Result<Vec<usize>, _>>()?;
```

File: beacon_node/rest_api/src/validator.rs (hunk 4 above)
Comment (spble) [reply beacon_node/rest_api/src/validator.rs]: I have added clarification to this, but the `EpochTooHigh` error will be returned in that case ,which will be included in the error message.

File: beacon_node/rest_api/src/helpers.rs (hunk 3 above)
Comment (spble) [reply beacon_node/rest_api/src/helpers.rs]: Yeah it does; I actually have the if statement around it that checks this commented out.
I will uncomment it.

**Comment by spble** [comment]
This PR has been merged into master.

**Pull request closed without merging**